src/content_scripts/applyColorSchemeJs.js

```diff
@@ -22,109 +51,294 @@ const ADDON_FAKED_WARNING = "matchMedia has been faked/overwritten by add-on web
  */
 function getColorTypeFromMediaQuery(mediaQueryString) {
     // to avoid expensive RegEx, first use a simple check
-    if (!mediaQueryString.includes(MEDIA_QUERY_COLOR_SCHEME)) {
+    if (!hasColorMediaQuery(mediaQueryString)) {
         return null;
     }
 
     if (MEDIA_QUERY_PREFER_COLOR[COLOR_STATUS.LIGHT].test(mediaQueryString)) {
         return COLOR_STATUS.LIGHT;
     } else if (MEDIA_QUERY_PREFER_COLOR[COLOR_STATUS.DARK].test(mediaQueryString)) {
         return COLOR_STATUS.DARK;
-    } else if (MEDIA_QUERY_PREFER_COLOR[COLOR_STATUS.NO_PREFERENCE]) {
+    } else if (MEDIA_QUERY_PREFER_COLOR[COLOR_STATUS.NO_PREFERENCE].test(mediaQueryString)) {
         return COLOR_STATUS.NO_PREFERENCE;
     } else {
         return null;
     }
 }
 
 /**
- * Returns a fake MediaQueryList as best as possible.
+ * Whether a string contains media query for color scheme
  *
  * @private
- * @param {string} mediaQueryString the original media query string
- * @param {COLOR_STATUS} askedFor
- * @returns {Object} looks like MediaQueryList
+ * @param {string} mediaQueryString
+ * @returns {boolean}
  */
-function fakeMediaQueryResult(mediaQueryString, askedFor) {
-    let matches = false;
+function hasColorMediaQuery(mediaQueryString) {
+    return (mediaQueryString.includes(MEDIA_QUERY_COLOR_SCHEME));
+}
 
-    // only return true, if the asked status is the same as the one we want to fake
-    if (fakedColorStatus === askedFor) {
-        matches = true;
+/**
+ * Evaluate a media query string.
+ * Returns null if the query has nothing to do with color.
+ *
+ * @private
+ * @param {string} mediaQueryString
+ * @returns {boolean|null}
+ */
+function evaluateMediaQuery(mediaQueryString) {
+    if (fakedColorStatus === COLOR_STATUS.NO_OVERWRITE) {
+        return matchMedia(mediaQueryString).matches;
     }
 
-    // get some parts of real object
-    const realQuery = matchMedia("(prefers-color-scheme: no-preference)");
+    let requestedMedia = getColorTypeFromMediaQuery(mediaQueryString);
+    if (requestedMedia === null) {
+        return null;
+    }
 
-    return {
-        media: mediaQueryString,
-        matches: matches,
-        onchange: realQuery.onchange || null, // real value was null anyway in my tests
-        addListener: (...args) => {
-            console.warn(ADDON_FAKED_WARNING);
+    return (fakedColorStatus === requestedMedia);
+}
 
-            return realQuery.addListener(...args);
+/**
+ * Keep track of listener addition
+ * Creates hook function if necessary
+ *
+ * @private
+ * @param {function} func original listener
+ * @param {MediaQueryList} mediaQueryList
+ * @param {boolean} isOnChange
+ * @returns {function} hook
+ */
+function trackOnListener(func, mediaQueryList, isOnChange) {
+    let entry = wmFuncToEntry.get(func);
```

Comment:
`wmFuncToEntry`  What? Is there any way to use a more descriptive (no abbreviations again) variable name? Clean coding and so on…